File: sdk/python/hitechcloud/resources/bare_metal.py
"""Bare Metal & Colocation resource (21 endpoints)"""

from .base import BaseResource

class BareMetalResource(BaseResource):
    """Bare Metal & Colocation: reinstall, diagnostics, rescue, server info, IPs, VLANs, power"""

    def list_reinstall_templates(self, service_id: int) -> dict:
        """GET /api/service/{id}/reinstall/templates - List OS templates and recipes"""
        return self._http.get(f"/api/service/{service_id}/reinstall/templates")

    def cancel_diagnostics(self, service_id: int) -> dict:
        """POST /api/service/{id}/diag/cancel - Cancel diagnostics operation"""
        return self._http.post(f"/api/service/{service_id}/diag/cancel")

    def list_diagnostics_templates(self, service_id: int) -> dict:
        """GET /api/service/{id}/diag/templates - List diagnostics templates"""
        return self._http.get(f"/api/service/{service_id}/diag/templates")

    def get_diagnostics_status(self, service_id: int) -> dict:
        """GET /api/service/{id}/diag - Show diagnostic operation status"""
        return self._http.get(f"/api/service/{service_id}/diag")

    def run_diagnostics(self, service_id: int, **kwargs) -> dict:
        """POST /api/service/{id}/diag - Start new diagnostic operation"""
        return self._http.post(f"/api/service/{service_id}/diag", kwargs)

    def list_rescue_templates(self, service_id: int) -> dict:
        """GET /api/service/{id}/rescue/templates - List rescue templates"""
        return self._http.get(f"/api/service/{service_id}/rescue/templates")

    def get_rescue_status(self, service_id: int) -> dict:
        """GET /api/service/{id}/rescue - Show rescue operation status"""
        return self._http.get(f"/api/service/{service_id}/rescue")

    def rescue(self, service_id: int, **kwargs) -> dict:
        """POST /api/service/{id}/rescue - Start rescue operation"""
        return self._http.post(f"/api/service/{service_id}/rescue", kwargs)

    def cancel_rescue(self, service_id: int) -> dict:
        """POST /api/service/{id}/rescue/cancel - Cancel rescue operation"""
        return self._http.post(f"/api/service/{service_id}/rescue/cancel")

    def get_server_info(self, service_id: int) -> dict:
        """GET /api/service/{id}/info - Get server hardware and OS details"""
        return self._http.get(f"/api/service/{service_id}/info")

    def update_hostname(self, service_id: int, hostname: str) -> dict:
        """POST /api/service/{id}/hostname - Update server hostname"""
        return self._http.post(f"/api/service/{service_id}/hostname", {"hostname": hostname})

    def list_ips(self, service_id: int) -> dict:
        """GET /api/service/{id}/ips - List server IPs"""
        return self._http.get(f"/api/service/{service_id}/ips")

    def add_ip(self, service_id: int, **kwargs) -> dict:
        """POST /api/service/{id}/ips - Add new IP to server"""
        return self._http.post(f"/api/service/{service_id}/ips", kwargs)

    def list_vlans(self, service_id: int) -> dict:
        """GET /api/service/{id}/vlans - List available VLANs"""
        return self._http.get(f"/api/service/{service_id}/vlans")

    def get_ip_details(self, service_id: int, ip: str) -> dict:
        """GET /api/service/{id}/ips/{ip} - Get IP details"""
        return self._http.get(f"/api/service/{service_id}/ips/{ip}")

    def edit_ip(self, service_id: int, ip: str, **kwargs) -> dict:
        """PUT /api/service/{id}/ips/{ip} - Edit IP"""
        return self._http.put(f"/api/service/{service_id}/ips/{ip}", kwargs)

    def get_reinstall_details(self, service_id: int) -> dict:
        """GET /api/service/{id}/reinstall - Return installation options"""
        return self._http.get(f"/api/service/{service_id}/reinstall")

    def reinstall(self, service_id: int, **kwargs) -> dict:
        """POST /api/service/{id}/reinstall - Reinstall server"""
        return self._http.post(f"/api/service/{service_id}/reinstall", kwargs)

    def get_status(self, service_id: int) -> dict:
        """GET /api/service/{id}/status - Return server power status"""
        return self._http.get(f"/api/service/{service_id}/status")

    def power_action(self, service_id: int, action: str) -> dict:
        """POST /api/service/{id}/power/{action} - Power action (start/stop/reboot/reset/shutdown)"""
        return self._http.post(f"/api/service/{service_id}/power/{action}")

    def list_servers_in_stock(self) -> dict:
        """GET /api/serverstock - List servers in stock"""
        return self._http.get("/api/serverstock")

File: sdk/ruby/lib/hitechcloud/resources/bare_metal.rb
module HiTechCloud
  module Resources
    class BareMetal < Base

        # Get reinstall templates
        def reinstall_templates(id)
          @http.get("/api/service/%{id}/reinstall/templates" % { id: id })
        end

        # Cancel diagnostics
        def cancel_diag(id)
          @http.post("/api/service/%{id}/diag/cancel" % { id: id })
        end

        # Get diag templates
        def diag_templates(id)
          @http.get("/api/service/%{id}/diag/templates" % { id: id })
        end

        # Get diagnostics
        def diag(id)
          @http.get("/api/service/%{id}/diag" % { id: id })
        end

        # Run diagnostics
        def run_diag(id, data)
          @http.post("/api/service/%{id}/diag" % { id: id }, data)
        end

        # Get rescue templates
        def rescue_templates(id)
          @http.get("/api/service/%{id}/rescue/templates" % { id: id })
        end

        # Get rescue mode
        def rescue(id)
          @http.get("/api/service/%{id}/rescue" % { id: id })
        end

        # Enter rescue mode
        def enter_rescue(id, data)
          @http.post("/api/service/%{id}/rescue" % { id: id }, data)
        end

        # Cancel rescue mode
        def cancel_rescue(id)
          @http.post("/api/service/%{id}/rescue/cancel" % { id: id })
        end

        # Get server info
        def info(id)
          @http.get("/api/service/%{id}/info" % { id: id })
        end

        # Set hostname
        def set_hostname(id, data)
          @http.post("/api/service/%{id}/hostname" % { id: id }, data)
        end

        # Get IPs
        def ips(id)
          @http.get("/api/service/%{id}/ips" % { id: id })
        end

        # Add IP
        def add_ip(id, data)
          @http.post("/api/service/%{id}/ips" % { id: id }, data)
        end

        # Get VLANs
        def vlans(id)
          @http.get("/api/service/%{id}/vlans" % { id: id })
        end

        # Get IP
        def ip(id, ip)
          @http.get("/api/service/%{id}/ips/%{ip}" % { id: id, ip: ip })
        end

        # Update IP
        def update_ip(id, ip, data)
          @http.put("/api/service/%{id}/ips/%{ip}" % { id: id, ip: ip }, data)
        end

        # Get reinstall info
        def reinstall(id)
          @http.get("/api/service/%{id}/reinstall" % { id: id })
        end

        # Reinstall
        def do_reinstall(id, data)
          @http.post("/api/service/%{id}/reinstall" % { id: id }, data)
        end

        # Get status
        def status(id)
          @http.get("/api/service/%{id}/status" % { id: id })
        end

        # Power action
        def power(id, action, data)
          @http.post("/api/service/%{id}/power/%{action}" % { id: id, action: action }, data)
        end

        # Get server stock
        def server_stock
          @http.get("/api/serverstock")
      end
    end
  end
end
